### PR TITLE
retries and backoff

### LINKED ISSUE #12: Addition of retries and backoff
Add feature of retries and backoff with its tests

### PR DESCRIPTION
**Pull request by asadali214**
[Master issue](https://github.com/apimatic/apimatic-codegen/issues/427) 

- addition retries and backoff configurations with setters
- retrying api call if enable retry configuration is turned on

Closes #12

### CODE REVIEW

File: src/Unirest/Request.php [hunk 1]
```diff
@@ -11,6 +13,14 @@ class Request
     private static $handle = null;
     private static $jsonOpts = array();
     private static $socketTimeout = null;
+    private static $enableRetries = false;            // should we enable retries feature
+    private static $maxNumberOfRetries = 3;           // total number of allowed retries
+    private static $retryOnTimeout = false;           // Should we retry on timeout?
+    private static $retryInterval = 1.0;              // Initial retry interval in seconds, to be increased by backoffFactor
+    private static $maxRetryInterval = 120;           // maximum retry wait time (commutative)
```
Comment (sufyankhanrao) [inline src/Unirest/Request.php]: The variable name is misleading. Name should be maximumRetryWaitTime.

Comment (asadali214) [reply src/Unirest/Request.php]: updated

**Review by sufyankhanrao — APPROVED** [review]
These changes look good to me.

File: src/Unirest/Request.php [hunk 2]
```diff
@@ -11,6 +13,14 @@ class Request
...
     private static $jsonOpts = array();
     private static $socketTimeout = null;
+    private static $enableRetries = false;            // should we enable retries feature
+    private static $maxNumberOfRetries = 3;           // total number of allowed retries
+    private static $retryOnTimeout = false;           // Should we retry on timeout?
+    private static $retryInterval = 1.0;              // Initial retry interval in seconds, to be increased by backoffFactor
+    private static $maximumRetryWaitTime = 120;           // maximum retry wait time (commutative)
+    private static $backoffFactor = 2.0;              // backoff factor to be used to increase retry interval
+    private static $httpStatusCodesToRetry = array(); // Http status codes to retry against
+    private static $httpMethodsToRetry = array();     // Http methods to retry against
```
Comment (thehappybug) [inline src/Unirest/Request.php]: Should these two settings have some sane default values? If someone enables retries but does not set the status code and methods to retry, a default sane behavior should kick in. I don't think that is the case right now in your code.

File: src/Unirest/Request.php [hunk 3]
```diff
@@ -471,9 +569,28 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
-        $response   = curl_exec(self::$handle);
-        $error      = curl_error(self::$handle);
-        $info       = self::getInfo();
+        $retryCount        = 0;                           // current retry count
+        $waitTime          = 0.0;                         // wait time in secs before current api call
+        $allowed_wait_time = self::$maximumRetryWaitTime; // remaining allowed wait time in seconds
+        do {
+            // If Retrying i.e. retryCount >= 1
+            if ($retryCount > 0) {
+                usleep($waitTime * 1000000);
```
Comment (thehappybug) [inline src/Unirest/Request.php]: Note: Values larger than 1000000 (i.e. sleeping for more than a second) may not be supported by the operating system. Conditionally use `sleep()` instead when the value is larger than 1000000.

Source: https://www.php.net/manual/en/function.usleep.php

File: src/Unirest/Request.php [hunk 4]
```diff
@@ -471,9 +569,28 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
+                // calculate remaining allowed wait Time
+                $allowed_wait_time -= $waitTime;
+            }
+
+            // Execution of api call
+            $response  = curl_exec(self::$handle);
```
Comment (thehappybug) [inline src/Unirest/Request.php:584]: Is it safe to retry Curl calls like this i.e reusing the same inputs? For example, what happens when someone passes a file in the request. Is the call able to read the file again and create another request in the next retry?

File: src/Unirest/Request.php [hunk 5]
```diff
@@ -488,6 +605,70 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
+        }
+        return 0;
+    }
+
+    public static function getResponseHeader($response, $info, $name)
+    {
+        $header_size = $info['header_size'];
+        $headers     = Response::parseHeaders(substr($response, 0, $header_size));
+
+        return key_exists($name, $headers) ? $headers[$name] : null;
```
Comment (thehappybug) [inline src/Unirest/Request.php]: Header name lookup should be case-insensitive.

File: src/Unirest/Response.php [hunk 6]
```diff
@@ -42,7 +42,7 @@ public function __construct($code, $raw_body, $headers, $json_args = array())
      * @param string $raw_headers raw headers
      * @return array
      */
-    private function parseHeaders($raw_headers)
+    public static function parseHeaders($raw_headers)
```
Comment (thehappybug) [inline src/Unirest/Response.php]: Is it necessary to make this public? The fewer public methods in this class, the better.

File: src/Unirest/Request.php [hunk 7]
```diff
@@ -488,6 +605,70 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
+            } else {
+                // if value is a date time string in format RFC1123
+                $retry_after_date = strtotime($retry_after);
+                return $retry_after_date == false ? 0 : $retry_after_date - time();
+            }
+        }
+        return 0;
+    }
+
+    public static function getResponseHeader($response, $info, $name)
```
Comment (thehappybug) [inline src/Unirest/Request.php]: Is it necessary to make this public? The fewer public methods in this class, the better.

File: src/Unirest/Request.php [hunk 8]
```diff
@@ -488,6 +605,70 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
+                if ($waitTime <= $allowed_wait_time) {
+                    // set retry wait time for next api call, only if its under allowed time
+                    $retryWaitTime = $waitTime;
+                }
+            }
+        }
+        return $retryWaitTime;
+    }
+
+    public static function getRetryAfterInSeconds($retry_after)
```
Comment (thehappybug) [inline src/Unirest/Request.php]: Is it necessary to make this public? The fewer public methods in this class, the better.

File: src/Unirest/Request.php [hunk 9]
```diff
@@ -488,6 +605,70 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
+     *
+     * @param $method             string|Method  HttpMethod of apiCall
+     * @param $response           string|bool    Response returned by server
+     * @param $info               mixed          Response info
+     * @param $error              string         Error returned by server
+     * @param $allowed_wait_time  int            Remaining allowed wait time
+     * @param $retryCount         int            Attempt number
+     * @return float  Wait time before sending the next apiCall
+     */
+    public static function getRetryWaitTime($method, $response, $info, $error, $allowed_wait_time, $retryCount)
```
Comment (thehappybug) [inline src/Unirest/Request.php]: Is it necessary to make this public? The fewer public methods in this class, the better.

File: src/Unirest/Request.php [hunk 10]
```diff
@@ -488,6 +605,70 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
+    }
+
+    public static function getRetryAfterInSeconds($retry_after)
+    {
+        if (isset($retry_after)) {
+            if (is_numeric($retry_after)) {
+                return (int)$retry_after; // if value is already in seconds
+            } else {
+                // if value is a date time string in format RFC1123
+                $retry_after_date = strtotime($retry_after);
```
Comment (thehappybug) [inline src/Unirest/Request.php]: `strtotime` is too flexible with the format of the date. Is there any method that parses only and only the Date format supported by the HTTP standard? It is better to error out than to support non-standard formats here and create a mismatch between different langauges' behaviors.

File: src/Unirest/Request.php [hunk 11]
```diff
@@ -488,6 +605,70 @@ public static function send($method, $url, $body = null, $headers = array(), $us
...
+                return $retry_after_date == false ? 0 : $retry_after_date - time();
+            }
+        }
+        return 0;
+    }
+
+    public static function getResponseHeader($response, $info, $name)
+    {
+        $header_size = $info['header_size'];
+        $headers     = Response::parseHeaders(substr($response, 0, $header_size));
```
Comment (thehappybug) [inline src/Unirest/Request.php]: Headers are being parsed twice now for each request; once when you call this method from the `getRetryWaitTime` method and then again in the [constructor of `Response` class](https://github.com/apimatic/unirest-php/blob/1fd06013d74aa3c9586094e0bc2396301cde5a37/src/Unirest/Response.php#L21). It is better to refactor the `Request` and `Response` classes to parse the headers only once and reuse the parsed form.

**Review by thehappybug — CHANGES_REQUESTED** [review]
One big design question I have is: why didn't you go for passing the retry settings per API call? Instead, I see static methods everywhere. This would have required that you add an optional `requestOptions` parameter to the `send` method that would let the caller pass in the options you have created per API call.

Apart from the request changes, I have some suggestions that you can address, maybe in a later release if you don't have time right now:
1. Methods that calculate retry values or fetch headers can be easily unit tested so we should at least write unit tests for them.
2. Have 1-2 integration tests that trigger the retry path as well.

**Review by thehappybug — CHANGES_REQUESTED** [review]
One big design question I have is: why didn't you go for passing the retry settings per API call? Instead, I see static methods everywhere. This would have required that you add an optional `requestOptions` parameter to the `send` method that would let the caller pass in the options you have created per API call.

Apart from the request changes, I have some suggestions that you can address, maybe in a later release if you don't have time right now:
1. Methods that calculate retry values or fetch headers can be easily unit tested so we should at least write unit tests for them.
2. Have 1-2 integration tests that trigger the retry path as well.

**Comment by asadali214** [comment]
> One big design question I have is: why didn't you go for passing the retry settings per API call? Instead, I see static methods everywhere.

Yes, for this design decision I have seen the retries feature implementation in C# and Java, there the retry configuration is passed only once into the client, and the client then uses those configurations for all API calls. So I have created static setters for all configurations, just as we have one already for setting timeout, basically my plan is to set these configurations only once and make sure the unirest-php won't change any of their values

**Comment by asadali214** [comment]
> 1. Methods that calculate retry values or fetch headers can be easily unit tested so we should at least write unit tests for them.
> 2. Have 1-2 integration tests that trigger the retry path as well.

Sure, @thehappybug we will do it after releasing retries in our sdks, for now this feature is being tested using our local api-tester and a spec containing retries' codegen settings and tests

File: src/Unirest/Request.php (hunk 11 above)
Comment (asadali214) [reply src/Unirest/Request.php]: Fixed

File: src/Unirest/Request.php (hunk 10 above)
Comment (asadali214) [reply src/Unirest/Request.php]: Fixed

File: src/Unirest/Request.php (hunk 9 above)
Comment (asadali214) [reply src/Unirest/Request.php]: Fixed

File: src/Unirest/Request.php (hunk 8 above)
Comment (asadali214) [reply src/Unirest/Request.php]: Fixed

File: src/Unirest/Request.php (hunk 7 above)
Comment (asadali214) [reply src/Unirest/Request.php]: Fixed

File: src/Unirest/Response.php (hunk 6 above)
Comment (asadali214) [reply src/Unirest/Response.php]: Fixed

File: src/Unirest/Request.php (hunk 5 above)
Comment (asadali214) [reply src/Unirest/Request.php]: Fixed

File: src/Unirest/Request.php (hunk 3 above)
Comment (asadali214) [reply src/Unirest/Request.php]: Fixed by using sleep with the combination of usleep

File: src/Unirest/Request.php (hunk 2 above)
Comment (asadali214) [reply src/Unirest/Request.php]: Fixed

File: src/Unirest/Request.php (hunk 4 above)
Comment (asadali214) [reply src/Unirest/Request.php:584]: Yes, its safe to retry calls while sending files, we have tested this with our tester files and multipart sdk

**Pull request merged**